darshan-util/pydarshan/darshan/backend/cffi_backend.py

```diff
@@ -657,3 +657,64 @@ def _log_get_heatmap_record(log):
     libdutil.darshan_free(buf[0])
     
     return rec
+
+
+def _df_to_rec(rec_dict, mod_name, rec_index_of_interest=None):
+    mod_type = _structdefs[mod_name]
+    counters_n_cols = rec_dict["counters"].shape[1]
+    fcounters_n_cols = rec_dict["fcounters"].shape[1]
+    size = (counters_n_cols +
+            fcounters_n_cols -
+            # id and rank columns are duplicated
+            # in counters and fcounters
+            2)
+    rbuf_orig = np.empty(shape=(size,), dtype=np.float64)
+    rbuf = ffi.cast(mod_type, ffi.from_buffer(rbuf_orig))
+    rbuf[0] = ffi.from_buffer(rbuf_orig)
+    for counter_name, dtype, cols in zip(["counters", "fcounters"],
+                                         [np.int64, np.float64],
+                                         [counters_n_cols, fcounters_n_cols]):
+        df = rec_dict[counter_name]
+        record = df.iloc[rec_index_of_interest, ...]
+        data = np.ascontiguousarray(record[:cols - 2].to_numpy(), dtype=dtype)
+        if counter_name == "fcounters":
+            rbuf[0].fcounters = data.tolist()
+        else:
+            rbuf[0].counters = data.tolist()
+    #rbuf[0].base_rec.rank = np.int64(record.loc["rank"])
+    #rbuf[0].base_rec.id = np.uint64(record.id)
```

Comment:
a careful review of the buffer memory layouts used here/elsewhere in this branch is probably the main thing needed to move this forward

Reply:
one thing we could do in testing is probably check the raw bytes--something like `np.frombuffer(ffi.buffer(rbuf[0])).tobytes()` vs. the original read-in bytes for a given record.